File: architecture-skills/design-patterns/examples/builder/builder.cpp
// Builder pattern in C++.
//
// The builder accumulates state and returns *this by reference for chaining;
// build() validates and returns an immutable product.

#include <iostream>
#include <map>
#include <optional>
#include <stdexcept>
#include <string>
#include <utility>

class HttpRequest {  // product: immutable after construction
public:
    HttpRequest(std::string method, std::string url,
                std::map<std::string, std::string> headers,
                std::optional<std::string> body, double timeoutSeconds)
        : method_(std::move(method)), url_(std::move(url)),
          headers_(std::move(headers)), body_(std::move(body)),
          timeoutSeconds_(timeoutSeconds) {}

    const std::string& method() const { return method_; }
    const std::string& url() const { return url_; }
    const std::map<std::string, std::string>& headers() const { return headers_; }
    const std::optional<std::string>& body() const { return body_; }
    double timeoutSeconds() const { return timeoutSeconds_; }

private:
    std::string method_;
    std::string url_;
    std::map<std::string, std::string> headers_;
    std::optional<std::string> body_;
    double timeoutSeconds_;
};

class HttpRequestBuilder {
public:
    HttpRequestBuilder(std::string method, std::string url)
        : method_(std::move(method)), url_(std::move(url)) {}

    HttpRequestBuilder& header(const std::string& name, const std::string& value) {
        headers_[name] = value;
        return *this;
    }

    HttpRequestBuilder& body(std::string body) {
        body_ = std::move(body);
        return *this;
    }

    HttpRequestBuilder& timeout(double seconds) {
        timeoutSeconds_ = seconds;
        return *this;
    }

    HttpRequest build() const {
        if (body_ && (method_ == "GET" || method_ == "HEAD")) {
            throw std::invalid_argument(method_ + " requests cannot have a body");
        }
        return HttpRequest(method_, url_, headers_, body_, timeoutSeconds_);
    }

private:
    std::string method_;
    std::string url_;
    std::map<std::string, std::string> headers_;
    std::optional<std::string> body_;
    double timeoutSeconds_ = 30.0;
};

int main() {
    HttpRequest request = HttpRequestBuilder("POST", "https://api.example.com/users")
                              .header("Content-Type", "application/json")
                              .header("Authorization", "Bearer token123")
                              .body(R"({"name": "Ada"})")
                              .timeout(5.0)
                              .build();

    std::cout << request.method() << ' ' << request.url()
              << " (timeout " << request.timeoutSeconds() << "s)\n";
    for (const auto& [name, value] : request.headers()) {
        std::cout << name << ": " << value << '\n';
    }
    if (request.body()) {
        std::cout << *request.body() << '\n';
    }
    return 0;
}

File: architecture-skills/design-patterns/examples/builder/builder.rs
// Builder pattern in Rust.
//
// The idiomatic way to construct complex structs (compare
// std::process::Command). The consuming builder (self by value) lets
// build() move accumulated state into the product without cloning.

use std::collections::BTreeMap;

/// Immutable product.
#[derive(Debug)]
pub struct HttpRequest {
    pub method: String,
    pub url: String,
    pub headers: BTreeMap<String, String>,
    pub body: Option<String>,
    pub timeout_seconds: f64,
}

pub struct HttpRequestBuilder {
    method: String,
    url: String,
    headers: BTreeMap<String, String>,
    body: Option<String>,
    timeout_seconds: f64,
}

impl HttpRequestBuilder {
    pub fn new(method: &str, url: &str) -> Self {
        Self {
            method: method.to_uppercase(),
            url: url.to_string(),
            headers: BTreeMap::new(),
            body: None,
            timeout_seconds: 30.0,
        }
    }

    pub fn header(mut self, name: &str, value: &str) -> Self {
        self.headers.insert(name.to_string(), value.to_string());
        self
    }

    pub fn body(mut self, body: &str) -> Self {
        self.body = Some(body.to_string());
        self
    }

    pub fn timeout(mut self, seconds: f64) -> Self {
        self.timeout_seconds = seconds;
        self
    }

    pub fn build(self) -> Result<HttpRequest, String> {
        if self.body.is_some() && matches!(self.method.as_str(), "GET" | "HEAD") {
            return Err(format!("{} requests cannot have a body", self.method));
        }
        Ok(HttpRequest {
            method: self.method,
            url: self.url,
            headers: self.headers,
            body: self.body,
            timeout_seconds: self.timeout_seconds,
        })
    }
}

fn main() {
    let request = HttpRequestBuilder::new("POST", "https://api.example.com/users")
        .header("Content-Type", "application/json")
        .header("Authorization", "Bearer token123")
        .body(r#"{"name": "Ada"}"#)
        .timeout(5.0)
        .build()
        .expect("valid request");

    println!("{request:#?}");
}
